src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+      r = inject_linkage(parent_ino, dname, fragment, dentry);
+      if (r < 0) {
+        return r;
+      }
+    }
+
+    if (!created_dirfrag) {
+      // If the parent dirfrag already existed, then stop traversing the
+      // backtrace: assume that the other ancestors already exist too.  This
+      // is an assumption rather than a truth, but it's a convenient way
+      // to avoid the risk of creating multiply-linked directories while
+      // injecting data.  If there are in fact missing ancestors, this
+      // should be fixed up using a separate tool scanning the metadata
+      // pool.
+      break;
+    } else {
+      // Proceed up the backtrace, creating parents
+      ino = parent_ino;
+    }
+
+    // TODO ensure that injected inode has layout pointing to this pool (if
+    // none of its ancestor layouts does)
+
+    // TODO handle backtraces pointing to stray dirs of MDS ranks that
+    // don't exist
+
+    // TODO handle strays in general: if something was stray but had
+    // hard links, we can't know its linked name, but we can shove it
+    // some recovery directory.
+
+    // TODO for objects with no backtrace, OPTIONALLY (don't do it by
+    // default) write them a /_recovered/<inodeno>.data backtrace and
+    // link them in there.  In general backtrace-less objects are
+    // just journalled metadata that hasn't landed yet, we should only
+    // override that if we are explicitly told to, or if a full
+    // forward scrub has failed to tag them.
+
+    // TODO scan objects for size of recovered inode.  We can either
+    // do this inline here, OR we can rely on a full scan also
+    // touching all other objects, and updating max size of inode
+    // every time we see it (but that gets complicated with multiple
+    // workers).  Maybe also a fast path for <4MB objects that sets size
+    // to the size of the 0000th object when that size is <4MB.  More generally,
+    // we could do this in an initial pass that just looks at all objects,
+    // and sets an xattr on the 0000th object (including creating it if necessary)
+    // to the maximum seen (may need a special RADOS op to do a "set if greater"
+    // xattr write)
+  }
+
+  return 0;
+}
+
+int MetadataDriver::find_or_create_dirfrag(
+    inodeno_t ino,
+    frag_t fragment,
+    bool *created)
+{
+  assert(created != NULL);
+
+  fnode_t existing_fnode;
+  *created = false;
+
+  int r = read_fnode(ino, fragment, &existing_fnode);
+  if (r == -ENOENT || r == -EINVAL) {
+    // Missing or corrupt fnode, create afresh
+    bufferlist fnode_bl;
+    fnode_t blank_fnode;
+    blank_fnode.version = 1;
+    blank_fnode.encode(fnode_bl);
+
+    // TODO: for the ENOENT case, do an 'exclusive create' op
+    // at the start of this write op, so that multiple
+    // writers won't both think they did the create.
+    // For the EINVAL case, we don't have a neat way to do
+    // that, so we will potentially have two writers thinking
```

Comment:
Actually, we can do this: grab the object, version out of the response, then compose a delete op that first asserts the version matches. If it passes, great, we're the winner; if it fails somebody else must have changed it.